docs/tutorial/codecs.txt

```diff
@@ -0,0 +1,433 @@
+       }
+
+       public function decode($value): Person
+       {
+           if (! $this->canDecode($value)) {
+               throw UnsupportedValueException::invalidDecodableValue($value);
+           }
+
+           $person = new Person($value->get('name'));
+           $person->id = $value->get('_id');
+
+           return $person;
+       }
+
+       public function encode($value): MongoDB\BSON\Document
+       {
+           if (! $this->canEncode($value)) {
+               throw UnsupportedValueException::invalidEncodableValue($value);
+           }
+
+           return MongoDB\BSON\Document::fromPHP([
+               '_id' => $value->id,
+               'name' => $value->name,
+           ]);
+       }
+   }
+
+To then use this codec with a collection, specify the ``codec`` option when selecting the collection:
+
+.. code-block:: php
+
+   <?php
+
+   $client = new MongoDB\Client();
+   $collection = $client->selectCollection('test', 'person', [
+       'codec' => new PersonCodec(),
+   ]);
+
+   $person = new Person('Jane Doe');
+   $collection->insertOne($person);
+
+   $person = $collection->findOne();
+
+The example above selects a collection and instructs it to use the ``PersonCodec`` for encoding and decoding documents.
+When inserting data, the ``PersonCodec`` is used to encode the document. When retrieving data, the ``PersonCodec`` is
+used to decode BSON data into a ``Person`` instance. Note that while the ``PersonCodec`` could technically decode any
+BSON document that contains a name field, we wouldn't want to use it for any other documents. Document codecs are meant
+to be used in a collection, or when decoding embedded documents.
+
+When working on a collection with a codec, the codec will only accept and return data of that type for certain
+operations. The ``bulkWrite``, ```findOneAndReplace``, ``insertMany``, ``insertOne``, and ``replaceOne`` operations
+will attempt to encode the given data using the provided codec. Trying to insert or replace a document that cannot be
+encoded will result in an exception. The ``aggregate``, ``find``, ``findOne``, ``findOneAndDelete``,
+``findOneAndReplace``, and ``findOneAndUpdate`` operations will attempt to decode returned documents using the provided
+codec. If the codec does not support the data returned, an exception will be thrown. You can disable codec usage for a
+specific operation or use a different codec (e.g. to decode the result of an aggregation pipeline) by specifying the
+nullable ``codec`` option for the operation. This will override the collection-level codec:
+
+.. code-block:: php
+
+   <?php
+
+   // Returns a document using the default type map
+   $collection->findOne(['name' => 'Jane Doe'], [
+       'codec' => null,
+   ]);
+
+   // Disables codec usage as the aggregate result will have a different format
+   $collection->aggregate([['$group' => [
+       '_id' => '$accountType',
+       'count' => ['$sum' => 1],
+   ]]], [
+       'codec' => null,
+   ]);
```

Comment:
This formatting looks unlike anything else I've found in our docs or project. A subsequent line with the same indentation as the first would usually be the final line in the statement.

Should this instead be:

```
$collection->aggregate(
    [['$group' => [
        ...
    ]]],
    ['codec' => null]
);
```

That honestly looks odd as well. I don't think there's any good way to do this -- so it's more about just picking something consistent. AFAIK this isn't addressed by our CS configuration.

Reply:
I've started formatting pipelines a bit different, even though phpcs will throw a fit about it:

```php
$collection->aggregate([
    ['$group' => [
        ...,
    ]],
]);
```

It does become a bit awkward with options, at which point I usually extract pipeline and options to their own variables to have cleaner looking invocations. That said, I think this would be a good compromise:
```php
$collection->aggregate(
    $pipeline,
    ['codec' => null],
);
```

We don't necessarily need to introduce a `$pipeline` variable in the same snippet as it's rather self-explanatory, and it helps keeping focus on the `codec` option.